rust/worker/src/execution/data/data_chunk.rs

```diff
@@ -0,0 +1,146 @@
+use std::ops::Range;
+use std::sync::Arc;
+
+use crate::types::EmbeddingRecord;
+
+#[derive(Clone, Debug)]
+pub(crate) struct DataChunk {
+    data: Arc<[EmbeddingRecord]>,
+    visibility: Arc<[bool]>,
+}
+
+impl DataChunk {
+    pub fn new(data: Arc<[EmbeddingRecord]>) -> Self {
+        let len = data.len();
+        DataChunk {
+            data,
+            visibility: vec![true; len].into(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn visible_len(&self) -> usize {
+        // return all the data with visibility true
+        self.visibility.iter().filter(|&v| *v).count()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    pub fn get(&self, index: usize) -> Option<&EmbeddingRecord> {
+        if index < self.data.len() {
+            Some(&self.data[index])
+        } else {
+            None
+        }
+    }
+
+    pub fn get_visibility(&self, index: usize) -> Option<bool> {
+        if index < self.visibility.len() {
+            Some(self.visibility[index])
+        } else {
+            None
+        }
+    }
+
+    pub fn iter(&self) -> DataChunkIteraror<'_> {
+        DataChunkIteraror {
+            chunk: self,
+            index: 0,
+        }
+    }
+
+    pub fn set_visibility(&mut self, visibility: Vec<bool>) {
```

Comment:
should this be indexed? I don't think callers should reallocate this memory? Instead I think we should design it such that we allocate visibility once and then callers can mutate the visibility at an index.